services/src/UseCase/Command/ChangeRoleForUser/Infrastructure/PersistenceRepository.php

```
@@ -0,0 +1,96 @@
+use App\UseCase\Command\ChangeRoleForUser\Repository;
+use DateTime;
+use Kununu\Doctrine\DBAL\Repository\AbstractRepository;
+
+final class PersistenceRepository extends AbstractRepository implements Repository
+{
+    public function findUserRoleByByUserUuidAndResource(UserUuid $userUuid, Resource $resource): UserRole
+    {
+        $params = $resource->toArray();
+        $params['user'] = $userUuid->toString();
+
+        $result = $this->connection
+            ->executeQuery($this->getQuery(GetUserRoleByUserUuidAndResource::class), $params)
+            ->fetch();
+
+        return UserRole::fromArray($result);
+    }
+
+    public function userHasRoleForResource(
+        UserUuid $userUuid,
+        string $resourceType,
+        ?string $resourceUuid = null
+    ): bool {
+        $params = [
+            $userUuid->toString(),
+            $resourceType,
+        ];
+        if ($resourceUuid) {
+            $params[] = $resourceUuid;
+        }
+
+        return (bool) $this->connection
+            ->executeQuery(
+                $this->getQuery(
+                    GetUserRolesByResource::class,
+                    false,
+                    $resourceUuid
+                ),
+                $params
+            )->rowCount();
+    }
+
+    public function findRoleId(RoleName $roleName): ?RoleId
+    {
+        $result = $this->connection->executeQuery(
+            $this->getQuery(GetRoleId::class),
+            [$roleName->value()]
+        )->fetchColumn();
+
+        return $result ? RoleId::fromInt((int) $result) : null;
+    }
+
+    public function changeRoleForUser(UserRole $userRole, RoleId $newRole, UserUuid $changer): void
+    {
+        $result = $this->connection
+            ->executeQuery(
+                $this->getQuery(UpdateRoleId::class),
+                [
+                    $newRole->toInt(),
+                    $userRole->id(),
+                ]
+            )->rowCount();
+
+        if ($result) {
```

If we cannot update !$result we should throw an exception no? In reality, nothing was changed

---

If we want the log to be persistent we should have this on a transaction

---

This check is just to make sure we don't insert anything if the update was not successful. In theory if there's an error during the update then some exception will be thrown.
Using the CQRS bundle, the command is already inside a transaction, don't think we need to add another transaction here